keras_nlp/models/gpt2.py

```diff
@@ -0,0 +1,287 @@
+                ),
+                layer_norm_epsilon=1e-05,
+                kernel_initializer=_gpt_2_kernel_initializer(stddev=0.02),
+                normalize_first=True,
+                name=f"transformer_layer_{i}",
+            )(x, decoder_padding_mask=padding_mask)
+
+        sequence_output = keras.layers.LayerNormalization(
+            name="layer_norm",
+            axis=-1,
+            epsilon=1e-05,
+            dtype=tf.float32,
+        )(x)
+
+        # Instantiate using Functional API Model constructor
+        super().__init__(
+            inputs={
+                "token_ids": token_ids,
+                "padding_mask": padding_mask,
+            },
+            outputs=sequence_output,
+            name=name,
+            trainable=trainable,
+        )
+        # All references to `self` below this line
+        self.vocabulary_size = vocabulary_size
+        self.num_layers = num_layers
+        self.num_heads = num_heads
+        self.hidden_dim = hidden_dim
+        self.intermediate_dim = intermediate_dim
+        self.dropout = dropout
+        self.max_sequence_length = max_sequence_length
+
+    def get_config(self):
+        config = super().get_config()
+        config.update(
+            {
+                "vocabulary_size": self.vocabulary_size,
+                "num_layers": self.num_layers,
+                "num_heads": self.num_heads,
+                "hidden_dim": self.hidden_dim,
+                "intermediate_dim": self.intermediate_dim,
+                "dropout": self.dropout,
+                "max_sequence_length": self.max_sequence_length,
+            }
+        )
+        return config
+
+
+MODEL_DOCSTRING = """GPT-2 implementation using "{type}"
+    architecture (with {num_params} parameters).
+
+    This network implements a Transformer-based decoder as
+    described in
+    ["Language Models are Unsupervised Multitask Learners"](https://cdn.openai.com/better-language-models/language_models_are_unsupervised_multitask_learners.pdf).
+    It includes the embedding lookups and transformer layers, but not the
+    language modeling or classification task heads.
+
+    Args:
+        vocabulary_size: int, optional. The size of the token vocabulary.
+        name: String, optional. Name of the model.
+        trainable: boolean, optional. If the model's variables should be
+            trainable.
+
+    Example usage:
+    ```python
+    # Randomly initialized Gpt2{type} encoder
+    model = keras_nlp.models.Gpt2{type}(vocabulary_size=10000)
+
+    # Call encoder on the inputs.
+    input_data = {{
+        "token_ids": tf.random.uniform(
+            shape=(1, 1024), dtype=tf.int64, maxval=model.vocabulary_size
+        ),
+        "padding_mask": tf.constant([1] * 1024, shape=(1, 1024)),
+    }}
+    output = model(input_data)
+"""
+
+
+def Gpt2Base(vocabulary_size, name=None, trainable=True):
+    return Gpt2Custom(
+        vocabulary_size=vocabulary_size,
```

Comment:
Is `vocabulary_size` actually fixed for all released checkpoints? I think so, but wanted to check my understanding

Reply:
Yep, fixed for all ckpts - 50257

https://openaipublic.blob.core.windows.net/gpt-2/models/355M/hparams.json, https://openaipublic.blob.core.windows.net/gpt-2/models/124M/hparams.json, etc.